proposals/0167-loader-v4.md

```diff
@@ -0,0 +1,435 @@
+    - the program account is writable, otherwise throw `InvalidArgument`
+    - the provided authority (instruction account at index 1) signed,
+    otherwise throw `MissingRequiredSignature`
+  - If this is not an initialization:
+    - Verify the program account
+    - Check that the status stored in the program account is
+    `NeverBeenDeployed` or `Retracted`,
+    otherwise throw `InvalidArgument`
+  - Check that there are enough funds in the program account for rent
+  exemption of the new length, otherwise throw `InsufficientFunds`
+  - Set the length of the program account to the requested new size plus
+  the header size
+  - In case that this is an initialization, also initialize the header:
+    - Set the slot to zero, **not** the current slot
+    - Set the authority address (from the instruction account at index 1)
+    - Set the status to `NeverBeenDeployed`
+
+#### WithdrawExcessLamports
+
+- Instruction accounts:
+  - `[writable]` The program account to withdraw from.
+  - `[signer]` The authority of the program.
+  - `[]` The recipient account.
+- Instruction data:
+  - Enum variant `3u32`
+- Behavior:
+  - Charge 32 CUs
+  - Check there are at least three instruction accounts,
+    otherwise throw `NotEnoughAccountKeys`
+  - Check that program account and recipient account do not alias,
+    otherwise throw `AccountBorrowFailed`
+  - Verify the program account
+  - Transfer lamports which are not needed for rent exemption from the
+  program account to the recipient account
+
+#### WithdrawAllLamports
+
+- Instruction accounts:
+  - `[writable]` The program account to withdraw from.
+  - `[signer]` The authority of the program.
+  - `[]` The recipient account.
+- Instruction data:
+  - Enum variant `4u32`
+- Behavior:
+  - Charge 32 CUs
+  - Check there are at least three instruction accounts,
+    otherwise throw `NotEnoughAccountKeys`
+  - Check that program account and recipient account do not alias,
+    otherwise throw `AccountBorrowFailed`
+  - Verify the program account
+  - Check that the status stored in the program account is
+  `NeverBeenDeployed` or `Retracted`,
+  otherwise throw `InvalidArgument`
+  - Set the length of the program account to 0 (removing the header too)
+  - Transfer all lamports from the program account to the recipient account
+
+#### Deploy
+
+- Instruction accounts:
+  - `[writable]` The program account to deploy.
+  - `[signer]` The authority of the program.
+- Instruction data:
+  - Enum variant `5u32`
+- Behavior:
+  - Charge 32 CUs
+  - Check there are at least two instruction accounts,
+    otherwise throw `NotEnoughAccountKeys`
+  - Verify the program account
+  - Check that the slot stored in the program account is not the current
+  (deployment cooldown), otherwise throw `InvalidArgument`
+  - Check that the status stored in the program account is `NeverBeenDeployed`
+  or `Retracted`
+    otherwise throw `InvalidArgument`
+  - Charge program_length_in_bytes CUs
```

Comment:
Accounts can be 10MB, but the max CUs for a transaction is 1.4M CUs. So all costs that use the program length will need to be reduced

Reply:
Good point, should probably divide by `cpi_bytes_per_unit`